File: leetcode/cpp/3.无重复字符的最长子串.cpp
/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
#include <string>
#include <unordered_set>
#include <algorithm>
using namespace std; 

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> set;
        int count(0);
        int r(0);
        for(int l=0;l<s.length();++l) {
            if(l > 0) {
                set.erase(s[l-1]);
            }
            while(r < s.length() && set.count(s[r]) == 0) {
                set.emplace(s[r]);
                r++;
            }
            count = max(count, r - l);
        }
        return count;
    }
};
// @lc code=end

File: leetcode/golang/3.无重复字符的最长子串.go
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	count := 0
	m := make(map[byte]int)

	r := 0
	for l := 0; l < len(s); l++ {
		if l > 0 {
			delete(m, s[l-1])
		}
		for r < len(s) && m[s[r]] == 0 {
			m[s[r]] = 1
			r++
		}
		count = max(count, r-l)
	}
	return count
}

func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

// @lc code=end

